File: solveBICGSTAB.py
import jax
import jax.numpy as jnp
import numpy as np
from jax.scipy.sparse.linalg import bicgstab
import timeit
import csv

gpu_device = jax.devices('gpu')[0]
cpu_device = jax.devices('cpu')[0]

def timeBICGSTAB(A, b):
  x = bicgstab(A, b)
  return x

dimensions = [10, 100, 1000, 2000]

elapsed_time_cpu = []
elapsed_time_gpu = []

for i, dimension in enumerate(dimensions):
    # loading data
    fileNameMatrixA = 'data/matrixA' + str(dimension) + '.csv'
    fileNameMatrixb = 'data/matrixb' + str(dimension) + '.csv'

    A = np.loadtxt(fileNameMatrixA, delimiter=',')
    b = np.loadtxt(fileNameMatrixb, delimiter=',')

    # converting to jax arrays
    A = jnp.array(A)
    b = jnp.array(b)

    with jax.default_device(cpu_device):
        start_time = timeit.default_timer()
        x_cpu = jax.jit(timeBICGSTAB)(A, b)
        end_time = timeit.default_timer()
        elapsed_time_cpu.append([i, (end_time - start_time)])


    with jax.default_device(gpu_device):
        start_time = timeit.default_timer()
        x_gpu = jax.jit(timeBICGSTAB)(A, b)
        end_time = timeit.default_timer()
        elapsed_time_gpu.append([i, (end_time - start_time)])

    print("\nCpu timing: ")
    print(elapsed_time_cpu[i][1])
    print("\nGpu timing: ")
    print(elapsed_time_gpu[i][1])

with open('timing_jax_cpu.csv', 'w', newline='') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerows(elapsed_time_cpu)

with open('timing_jax_gpu.csv', 'w', newline='') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerows(elapsed_time_gpu)

File: solveBICGSTAB.jl
using DelimitedFiles, IterativeSolvers, SparseArrays, LinearAlgebra, CSV, Tables

function timeBICGSTAB(A, b, l)
    x = bicgstabl(A, b, l, reltol=1.0e-5)
    return x
end

dimensions = [10, 100, 1000, 2000]

println(dimensions)

l = 2
timing = zeros(Float16, length(dimensions), 2)
for (i, dimension) in enumerate(dimensions)
    println(i)
    A = readdlm("data/matrixA" * string(dimension) * ".csv", ',', Float64, '\n')
    b = readdlm("data/matrixb" * string(dimension) * ".csv", ',', Float64, '\n')
    timing[i, :] = [dimension, @elapsed timeBICGSTAB(A, b, l)]

    println("Time: ", timing[i, 2])
end

CSV.write("timing_Julia.csv", Tables.table(timing), writeheader=false)